File: solver/electrostatic/electrostatic_solver.py
"""The electrostatic solver solves for the electric potential and the electric
field only.
"""

import numpy as np
import scipy.sparse
from proto.solver_config_pb2 import SolverConfig

from mesh.gmsh_interface import GmshNodeType
from model.material import MaterialProperties
from solver.electromagnetic_solver import (ElectromagneticSolver,
                                           ElectromagneticSolver2D,
                                           ElectromagneticSolver3D)
from solver.neighbor_lookup import NeighborLookup


class ElectrostaticSolver(ElectromagneticSolver):
    """Interface for an electrostatic solver."""

    def __init__(self, mesh_file: str, solver_config: SolverConfig) -> None:
        super().__init__(mesh_file, solver_config)

    @property
    def num_unknowns(self) -> int:
        """Returns the number of unknowns."""
        return (self.num_electric_potential_unknowns +
                self.num_electric_field_unknowns)

    def _get_dc_voltage(self, tag: int) -> float:
        """Returns the DC voltage for the tag.

        Args:
            tag: Tag of the entity.

        Raises:
            ValueError: If the tag cannot be found.
        """
        for entity_config in self.config.entity_configs:
            if entity_config.tag == tag:
                return entity_config.dc_voltage
        raise ValueError(f"Entity {tag} cannot be found.")


class ElectrostaticSolver2D(ElectrostaticSolver, ElectromagneticSolver2D):
    """Interface for a 2D electrostatic solver."""

    def _solve(self) -> None:
        """Implementation for solving for the electric potential, the electric
        field, the magnetic vector potential, and the magnetic flux density.

        In the matrix-vector equation, the first unknowns correspond to the
        nodes' electric potentials, and the remaining unknowns correspond to
        the nodes' electric fields.
        Similarly, the first equations correspond to Poisson's equation and any
        voltage boundary conditions, and the remaining equations correspond to
        the electric field equations for each node in the x and y-directions.
        """
        # Get the node coordinates.
        node_tags, node_coordinates = self.get_node_coordinates(
            dim=self.dimension(), coordinates_dim=self.dimension())
        node_tag_to_coordinates = dict(zip(node_tags, node_coordinates))

        # Get the conductor and the insulator entities.
        physical_groups = self.get_physical_groups(dim=self.dimension())
        conductor_entity_tags = []
        insulator_entity_tags = []
        for _, physical_group_tag in physical_groups:
            material = self.get_material_for_physical_group(
                tag=physical_group_tag)
            entities = self.get_entities_for_physical_group(
                dim=self.dimension(), tag=physical_group_tag)
            if MaterialProperties.is_conductor(material):
                conductor_entity_tags.extend(entities)
            elif MaterialProperties.is_insulator(material):
                insulator_entity_tags.extend(entities)

        # Initialize the matrix-vector equation.
        A = scipy.sparse.lil_matrix((self.num_unknowns, self.num_unknowns),
                                    dtype=np.float64)
        b = np.zeros(self.num_unknowns)

        # Fill in Poisson's equations and the electric field equations for the
        # nodes within the insulators, including the boundary nodes.
        for insulator_tag in insulator_entity_tags:
            _, insulator_triangle_node_tags = self.get_faces(tag=insulator_tag)
            insulator_triangle_neighbors = NeighborLookup(
                insulator_triangle_node_tags)
            insulator_node_tags = self.get_nodes(tag=insulator_tag,
                                                 dim=self.dimension())
            for tag in insulator_node_tags:
                x, y = node_tag_to_coordinates[tag]
                poisson_equation_index = (
                    self._get_poisson_electric_potential_equation_index(tag))
                electric_field_x_equation_index = (
                    self._get_electric_field_x_equation_index(tag))
                electric_field_y_equation_index = (
                    self._get_electric_field_y_equation_index(tag))
                electric_potential_unknown_index = (
                    self._get_electric_potential_unknown_index(tag))
                electric_field_x_unknown_index = (
                    self._get_electric_field_x_unknown_index(tag))
                electric_field_y_unknown_index = (
                    self._get_electric_field_y_unknown_index(tag))
                num_adjacent_triangles = (
                    insulator_triangle_neighbors.get_num_adjacent_entities(tag))

                # Iterate over all adjacent triangles.
                for (tag_neighbor1, tag_neighbor2
                    ) in insulator_triangle_neighbors.get_neighbors(tag):
                    x_neighbor1, y_neighbor1 = node_tag_to_coordinates[
                        tag_neighbor1]
                    x_neighbor2, y_neighbor2 = node_tag_to_coordinates[
                        tag_neighbor2]
                    electric_potential_unknown_index_neighbor1 = (
                        self._get_electric_potential_unknown_index(
                            tag_neighbor1))
                    electric_potential_unknown_index_neighbor2 = (
                        self._get_electric_potential_unknown_index(
                            tag_neighbor2))
                    electric_field_x_unknown_index_neighbor1 = (
                        self._get_electric_field_x_unknown_index(tag_neighbor1))
                    electric_field_y_unknown_index_neighbor1 = (
                        self._get_electric_field_y_unknown_index(tag_neighbor1))
                    electric_field_x_unknown_index_neighbor2 = (
                        self._get_electric_field_x_unknown_index(tag_neighbor2))
                    electric_field_y_unknown_index_neighbor2 = (
                        self._get_electric_field_y_unknown_index(tag_neighbor2))

                    # Calculate the denominator.
                    denominator = (x_neighbor1 * y_neighbor2 - x_neighbor1 * y -
                                   x_neighbor2 * y_neighbor1 + x_neighbor2 * y +
                                   x * y_neighbor1 - x * y_neighbor2)

                    # Add the coefficients of the electric fields for Poisson's
                    # equation.
                    A[poisson_equation_index,
                      electric_field_x_unknown_index_neighbor1] += (
                          (y_neighbor2 - y) / denominator)
                    A[poisson_equation_index,
                      electric_field_y_unknown_index_neighbor1] += (
                          (x - x_neighbor2) / denominator)
                    A[poisson_equation_index,
                      electric_field_x_unknown_index_neighbor2] += (
                          (y - y_neighbor1) / denominator)
                    A[poisson_equation_index,
                      electric_field_y_unknown_index_neighbor2] += (
                          (x_neighbor1 - x) / denominator)
                    A[poisson_equation_index,
                      electric_field_x_unknown_index] += (
                          (y_neighbor1 - y_neighbor2) / denominator)
                    A[poisson_equation_index,
                      electric_field_y_unknown_index] += (
                          (x_neighbor2 - x_neighbor1) / denominator)

                    # Add the coefficients of the electric potentials for the
                    # electric field equation in the x-direction.
                    A[electric_field_x_equation_index,
                      electric_potential_unknown_index_neighbor1] += (
                          (y_neighbor2 - y) / denominator)
                    A[electric_field_x_equation_index,
                      electric_potential_unknown_index_neighbor2] += (
                          (y - y_neighbor1) / denominator)
                    A[electric_field_x_equation_index,
                      electric_potential_unknown_index] += (
                          (y_neighbor1 - y_neighbor2) / denominator)

                    # Add the coefficients of the electric potentials for the
                    # electric field equation in the y-direction.
                    A[electric_field_y_equation_index,
                      electric_potential_unknown_index_neighbor1] += (
                          (x - x_neighbor2) / denominator)
                    A[electric_field_y_equation_index,
                      electric_potential_unknown_index_neighbor2] += (
                          (x_neighbor1 - x) / denominator)
                    A[electric_field_y_equation_index,
                      electric_potential_unknown_index] += (
                          (x_neighbor2 - x_neighbor1) / denominator)

                # Set the coefficient for the electric fields for the
                # electric field equations.
                A[electric_field_x_equation_index,
                  electric_field_x_unknown_index] = num_adjacent_triangles
                A[electric_field_y_equation_index,
                  electric_field_y_unknown_index] = num_adjacent_triangles

        # Fill in the electric potential and electric field equations for the
        # nodes within the conductors, revising the equations for the boundary
        # nodes.
        # At steady state, the electric potential is constant throughout the
        # insulator, and the electric field is zero throughout, even with a
        # non-zero resistivity.
        for conductor_tag in conductor_entity_tags:
            # Set the voltage boundary conditions and electric fields within
            # the capacitor plates.
            conductor_internal_node_tags = self.get_nodes(
                tag=conductor_tag,
                dim=self.dimension(),
                node_type=GmshNodeType.INTERNAL)
            for tag in conductor_internal_node_tags:
                electric_potential_unknown_index = (
                    self._get_electric_potential_unknown_index(tag))
                electric_field_x_unknown_index = (
                    self._get_electric_field_x_unknown_index(tag))
                electric_field_y_unknown_index = (
                    self._get_electric_field_y_unknown_index(tag))

                # Set the voltage boundary conditions.
                electric_potential_equation_index = (
                    self._get_poisson_electric_potential_equation_index(tag))
                A[electric_potential_equation_index,
                  electric_potential_unknown_index] = 1
                b[electric_potential_equation_index] = (
                    self._get_dc_voltage(conductor_tag))

                # Set the electric field boundary conditions in the x-direction.
                electric_field_x_equation_index = (
                    self._get_electric_field_x_equation_index(tag))
                A[electric_field_x_equation_index,
                  electric_field_x_unknown_index] = 1

                # Set the electric field boundary conditions in the y-direction.
                electric_field_y_equation_index = (
                    self._get_electric_field_y_equation_index(tag))
                A[electric_field_y_equation_index,
                  electric_field_y_unknown_index] = 1

            # Set the voltage boundary conditions for the boundary nodes at the
            # capacitor plates.
            conductor_boundary_node_tags = self.get_nodes(
                tag=conductor_tag,
                dim=self.dimension(),
                node_type=GmshNodeType.BOUNDARY)
            for tag in conductor_boundary_node_tags:
                electric_potential_unknown_index = (
                    self._get_electric_potential_unknown_index(tag))

                # Set the voltage boundary conditions.
                electric_potential_equation_index = (
                    self._get_poisson_electric_potential_equation_index(tag))
                A[electric_potential_equation_index, :] = 0
                A[electric_potential_equation_index,
                  electric_potential_unknown_index] = 1
                b[electric_potential_equation_index] = (
                    self._get_dc_voltage(conductor_tag))

        # Solve for the electric potential and the electric field.
        x = scipy.sparse.linalg.spsolve(A.tocsr(), b)
        self.electric_potential = x[:self.num_electric_potential_unknowns]
        self.electric_field = np.reshape(
            x[self.num_electric_potential_unknowns:self.
              num_electric_potential_unknowns +
              self.num_electric_field_unknowns], (-1, self.dimension()))

    def _get_electric_potential_unknown_index(self, tag: int) -> int:
        """Returns the unknown index corresponding to the node's electric
        potential.
        """
        return self._get_index_from_tag(tag)

    def _get_electric_field_x_unknown_index(self, tag: int) -> int:
        """Returns the unknown index corresponding to the node's electric field
        in the x-direction.
        """
        return self.dimension() * self._get_index_from_tag(
            tag) + self.num_electric_potential_unknowns

    def _get_electric_field_y_unknown_index(self, tag: int) -> int:
        """Returns the unknown index corresponding to the node's electric field
        in the y-direction.
        """
        return self.dimension() * self._get_index_from_tag(
            tag) + 1 + self.num_electric_potential_unknowns

    def _get_poisson_electric_potential_equation_index(self, tag: int) -> int:
        """Returns the equation index corresponding to Poisson's equation or
        any voltage boundary condition.
        """
        return self._get_index_from_tag(tag)

    def _get_electric_field_x_equation_index(self, tag: int) -> int:
        """Returns the equation index corresponding to the node's electric field
        in the x-direction.
        """
        return self.dimension() * self._get_index_from_tag(
            tag) + self.num_electric_potential_unknowns

    def _get_electric_field_y_equation_index(self, tag: int) -> int:
        """Returns the equation index corresponding to the node's electric field
        in the y-direction.
        """
        return self.dimension() * self._get_index_from_tag(
            tag) + 1 + self.num_electric_potential_unknowns


class ElectrostaticSolver3D(ElectrostaticSolver, ElectromagneticSolver3D):
    """Interface for a 3D electrostatic solver."""

    pass

File: solver/electrostatic/electrostatic_solver.cc
#include "solver/electrostatic/electrostatic_solver.h"

#include <Eigen/Dense>
#include <Eigen/SparseCore>
#include <Eigen/SparseLU>
#include <complex>
#include <forward_list>
#include <stdexcept>
#include <unordered_set>

#include "absl/strings/str_format.h"
#include "mesh/gmsh_interface.h"
#include "model/material.h"
#include "solver/neighbor_lookup.h"

namespace solver {

template <std::size_t Dimension>
double ElectrostaticSolver<Dimension>::GetDcVoltage(const int tag) {
  for (const auto& entity_config : this->config_.entity_configs()) {
    if (entity_config.tag() == tag) {
      return entity_config.dc_voltage();
    }
  }
  throw std::invalid_argument(
      absl::StrFormat("Entity %d cannot be found.", tag));
}

void ElectrostaticSolver2D::SolveImpl() {
  // In the matrix-vector equation, the first unknowns correspond to the nodes'
  // electric potentials, and the remaining unknowns correspond to the nodes'
  // electric fields.
  // Similarly, the first equations correspond to Poisson's equation and any
  // voltage boundary conditions, and the remaining equations correspond to the
  // electric field equations for each node in the x and y-directions.

  // Get the node coordinates.
  const auto node_tag_to_coordinates =
      GetNodeCoordinates(/*tag=*/-1, dimension());

  // Get the conductor and the insulator entities.
  const auto physical_groups = GetPhysicalGroups(dimension());
  std::unordered_set<int> conductor_entity_tags;
  std::unordered_set<int> insulator_entity_tags;
  for (const auto& [physical_group_dimension, physical_group_tag] :
       physical_groups) {
    const auto material = GetMaterialForPhysicalGroup(physical_group_tag);
    const auto& material_properties =
        model::MaterialProperties::kMaterialToProperties.at(material);
    const auto entities =
        GetEntitiesForPhysicalGroups(dimension(), physical_group_tag);
    if (material_properties.is_conductor()) {
      conductor_entity_tags.insert(entities.cbegin(), entities.cend());
    } else if (material_properties.is_insulator()) {
      insulator_entity_tags.insert(entities.cbegin(), entities.cend());
    }
  }

  // Initialize the matrix-vector triplets.
  std::forward_list<Eigen::Triplet<std::complex<double>>> A_triplets;
  std::forward_list<Eigen::Triplet<std::complex<double>>> b_triplets;

  // Fill in the electric potential and electric field equations for the nodes
  // within the conductors. For the boundary nodes, only fill in the electric
  // potential equations.
  // At steady state, the electric potential is constant throughout the
  // insulator, and the electric field is zero throughout, even with a non-zero
  // resistivity.
  std::unordered_set<gmsh::Tag> boundary_tags;
  for (const auto conductor_tag : conductor_entity_tags) {
    // Set the voltage boundary conditions and electric fields within the
    // conductor.
    const auto conductor_internal_node_tags =
        GetNodes(conductor_tag, dimension(), gmsh::NodeType::kInternal);
    for (const auto tag : conductor_internal_node_tags) {
      // Equation indices.
      const auto poisson_electric_potential_equation_index =
          this->poisson_electric_potential_equation_index(tag);
      const auto electric_field_x_equation_index =
          this->electric_field_x_equation_index(tag);
      const auto electric_field_y_equation_index =
          this->electric_field_y_equation_index(tag);

      // Unknown indices.
      const auto electric_potential_unknown_index =
          this->electric_potential_unknown_index(tag);
      const auto electric_field_x_unknown_index =
          this->electric_field_x_unknown_index(tag);
      const auto electric_field_y_unknown_index =
          this->electric_field_y_unknown_index(tag);

      // Set the voltage boundary conditions.
      A_triplets.emplace_front(poisson_electric_potential_equation_index,
                               electric_potential_unknown_index, 1);
      b_triplets.emplace_front(poisson_electric_potential_equation_index, 0,
                               GetDcVoltage(conductor_tag));

      // Set the electric field boundary conditions.
      A_triplets.emplace_front(electric_field_x_equation_index,
                               electric_field_x_unknown_index, 1);
      A_triplets.emplace_front(electric_field_y_equation_index,
                               electric_field_y_unknown_index, 1);
    }

    // Set the voltage boundary conditions for the boundary nodes of the
    // conductor.
    const auto conductor_boundary_node_tags =
        GetNodes(conductor_tag, dimension(), gmsh::NodeType::kBoundary);
    for (const auto tag : conductor_boundary_node_tags) {
      // Equation indices.
      const auto poisson_electric_potential_equation_index =
          this->poisson_electric_potential_equation_index(tag);

      // Unknown indices.
      const auto electric_potential_unknown_index =
          this->electric_potential_unknown_index(tag);

      // Set the voltage boundary conditions.
      A_triplets.emplace_front(poisson_electric_potential_equation_index,
                               electric_potential_unknown_index, 1);
      b_triplets.emplace_front(poisson_electric_potential_equation_index, 0,
                               GetDcVoltage(conductor_tag));
    }
    boundary_tags.insert(conductor_boundary_node_tags.cbegin(),
                         conductor_boundary_node_tags.cend());
  }

  // Fill in Poisson's equations and the electric field equations for the nodes
  // within the insulators, including the boundary nodes.
  for (const auto insulator_tag : insulator_entity_tags) {
    const auto insulator_triangles = GetFaces(insulator_tag);
    const auto insulator_triangle_neighbors =
        NeighborLookup2D(insulator_triangles);
    const auto insulator_node_tags =
        GetNodes(insulator_tag, dimension(), gmsh::NodeType::kAll);
    for (const auto tag : insulator_node_tags) {
      const auto& node_coordinates = node_tag_to_coordinates.at(tag);
      const auto x = node_coordinates[0];
      const auto y = node_coordinates[1];
      const auto num_adjacent_triangles =
          insulator_triangle_neighbors.num_adjacent_entities(tag);

      // Equation indices.
      const auto poisson_electric_potential_equation_index =
          this->poisson_electric_potential_equation_index(tag);
      const auto electric_field_x_equation_index =
          this->electric_field_x_equation_index(tag);
      const auto electric_field_y_equation_index =
          this->electric_field_y_equation_index(tag);

      // Unknown indices.
      const auto electric_potential_unknown_index =
          this->electric_potential_unknown_index(tag);
      const auto electric_field_x_unknown_index =
          this->electric_field_x_unknown_index(tag);
      const auto electric_field_y_unknown_index =
          this->electric_field_y_unknown_index(tag);

      // Iterate over all adjacent triangles.
      for (const auto& [tag_neighbor1, tag_neighbor2] :
           insulator_triangle_neighbors.neighbors(tag)) {
        const auto& node_coordinates_neighbor1 =
            node_tag_to_coordinates.at(tag_neighbor1);
        const auto x_neighbor1 = node_coordinates_neighbor1[0];
        const auto y_neighbor1 = node_coordinates_neighbor1[1];
        const auto& node_coordinates_neighbor2 =
            node_tag_to_coordinates.at(tag_neighbor2);
        const auto x_neighbor2 = node_coordinates_neighbor2[0];
        const auto y_neighbor2 = node_coordinates_neighbor2[1];

        // Unknown indices for both neighbors.
        const auto electric_potential_unknown_index_neighbor1 =
            this->electric_potential_unknown_index(tag_neighbor1);
        const auto electric_potential_unknown_index_neighbor2 =
            this->electric_potential_unknown_index(tag_neighbor2);
        const auto electric_field_x_unknown_index_neighbor1 =
            this->electric_field_x_unknown_index(tag_neighbor1);
        const auto electric_field_x_unknown_index_neighbor2 =
            this->electric_field_x_unknown_index(tag_neighbor2);
        const auto electric_field_y_unknown_index_neighbor1 =
            this->electric_field_y_unknown_index(tag_neighbor1);
        const auto electric_field_y_unknown_index_neighbor2 =
            this->electric_field_y_unknown_index(tag_neighbor2);

        // Calculate the denominator.
        const auto denominator = (x_neighbor1 * y_neighbor2 - x_neighbor1 * y -
                                  x_neighbor2 * y_neighbor1 + x_neighbor2 * y +
                                  x * y_neighbor1 - x * y_neighbor2);

        // Add the coefficients of the electric fields for Poisson's equation.
        // Boundary nodes already have a voltage boundary condition, so skip
        // boundary nodes.
        if (!boundary_tags.contains(tag)) {
          A_triplets.emplace_front(poisson_electric_potential_equation_index,
                                   electric_field_x_unknown_index_neighbor1,
                                   (y_neighbor2 - y) / denominator);
          A_triplets.emplace_front(poisson_electric_potential_equation_index,
                                   electric_field_y_unknown_index_neighbor1,
                                   (x - x_neighbor2) / denominator);
          A_triplets.emplace_front(poisson_electric_potential_equation_index,
                                   electric_field_x_unknown_index_neighbor2,
                                   (y - y_neighbor1) / denominator);
          A_triplets.emplace_front(poisson_electric_potential_equation_index,
                                   electric_field_y_unknown_index_neighbor2,
                                   (x_neighbor1 - x) / denominator);
          A_triplets.emplace_front(poisson_electric_potential_equation_index,
                                   electric_field_x_unknown_index,
                                   (y_neighbor1 - y_neighbor2) / denominator);
          A_triplets.emplace_front(poisson_electric_potential_equation_index,
                                   electric_field_y_unknown_index,
                                   (x_neighbor2 - x_neighbor1) / denominator);
        }

        // Add the coefficients of the electric potentials for the electric
        // field equation in the x-direction.
        A_triplets.emplace_front(electric_field_x_equation_index,
                                 electric_potential_unknown_index_neighbor1,
                                 (y_neighbor2 - y) / denominator);
        A_triplets.emplace_front(electric_field_x_equation_index,
                                 electric_potential_unknown_index_neighbor2,
                                 (y - y_neighbor1) / denominator);
        A_triplets.emplace_front(electric_field_x_equation_index,
                                 electric_potential_unknown_index,
                                 (y_neighbor1 - y_neighbor2) / denominator);

        // Add the coefficients of the electric potentials for the electric
        // field equation in the y-direction.
        A_triplets.emplace_front(electric_field_y_equation_index,
                                 electric_potential_unknown_index_neighbor1,
                                 (x - x_neighbor2) / denominator);
        A_triplets.emplace_front(electric_field_y_equation_index,
                                 electric_potential_unknown_index_neighbor2,
                                 (x_neighbor1 - x) / denominator);
        A_triplets.emplace_front(electric_field_y_equation_index,
                                 electric_potential_unknown_index,
                                 (x_neighbor2 - x_neighbor1) / denominator);
      }

      // Set the coefficient for the electric fields for the electric field
      // equations.
      A_triplets.emplace_front(electric_field_x_equation_index,
                               electric_field_x_unknown_index,
                               num_adjacent_triangles);
      A_triplets.emplace_front(electric_field_y_equation_index,
                               electric_field_y_unknown_index,
                               num_adjacent_triangles);
    }
  }

  // Create the sparse matrix-vector equation.
  Eigen::SparseMatrix<std::complex<double>> A(num_unknowns(), num_unknowns());
  A.setFromTriplets(A_triplets.cbegin(), A_triplets.cend());
  Eigen::SparseMatrix<std::complex<double>> b(num_unknowns(), 1);
  b.setFromTriplets(b_triplets.cbegin(), b_triplets.cend());

  // Solve for the electric potential and the electric field.
  Eigen::SparseLU<Eigen::SparseMatrix<std::complex<double>>> solver;
  solver.compute(A);
  if (solver.info() != Eigen::Success) {
    throw std::runtime_error(absl::StrFormat(
        "Failed to compute the decomposition of the matrix: %d.",
        solver.info()));
  }
  Eigen::VectorXcd x = solver.solve(b);
  electric_potential_ = x(Eigen::seqN(0, num_electric_potential_unknowns()));
  electric_field_ = x(Eigen::seqN(num_electric_potential_unknowns(),
                                  num_electric_field_unknowns()))
                        .reshaped<Eigen::RowMajor>(num_nodes_, dimension());
}

}  // namespace solver
